engineapi/test_crafting.py

```diff
@@ -0,0 +1,1022 @@
+
+        cls.crafting = CraftingFacet.CraftingFacet(cls.diamond_address)
+
+    def _create_recipe(
+        self,
+        recipe: Recipe,
+    ) -> int:
+        self.crafting.add_recipe(recipe.to_tuple(), {"from": accounts[0]})
+        recipe_id = self.crafting.num_recipes()
+        return recipe_id
+
+    def test_create_craft(self):
+        inputs = [
+            CraftingInput(
+                20,
+                self.erc20_contracts[1].address,
+                0,
+                10 * 10 ** 18,
+                0,
+            ),
+        ]
+        outputs = [
+            CraftingOutput(
+                20,
+                self.erc20_contracts[2].address,
+                0,
+                10 * 10 ** 18,
+                0,
+            ),
+        ]
+        num_recipes_before = self.crafting.num_recipes()
+        recipe = Recipe(inputs, outputs, True)
+        recipe_id = self._create_recipe(recipe)
+        num_recipes_after = self.crafting.num_recipes()
+
+        self.assertEqual(num_recipes_after, num_recipes_before + 1)
+
+        recipe_from_contract = self.crafting.get_recipe(recipe_id)
+        self.assertEqual(recipe_from_contract, recipe.to_tuple())
+
+    def _check_balances_after_crafting(self, recipe: Recipe, block_no_before: int):
+        for input in recipe.inputs:
+            (
+                balance_user_before,
+                balance_user_after,
+                balance_contract_before,
+                balance_contract_after,
+            ) = (None, None, None, None)
+            if input.token_type == 20:
+                erc20_contract = MockErc20.MockErc20(input.token_address)
+                balance_user_before = erc20_contract.balance_of(
+                    accounts[1], block_number=block_no_before
+                )
+                balance_user_after = erc20_contract.balance_of(accounts[1])
+                balance_contract_before = erc20_contract.balance_of(
+                    self.crafting.address, block_number=block_no_before
+                )
+                balance_contract_after = erc20_contract.balance_of(
+                    self.crafting.address
+                )
+            elif input.token_type == 1155:
+                erc1155_contract = MockERC1155.MockERC1155(input.token_address)
+                balance_user_before = erc1155_contract.balance_of(
+                    accounts[1], input.token_id, block_number=block_no_before
+                )
+                balance_user_after = erc1155_contract.balance_of(
+                    accounts[1], input.token_id
+                )
+                balance_contract_before = erc1155_contract.balance_of(
+                    self.crafting.address, input.token_id, block_number=block_no_before
+                )
+                balance_contract_after = erc1155_contract.balance_of(
+                    self.crafting.address,
+                    input.token_id,
+                )
+
+            if input.token_action == 0:
```

Comment:
This is where it would help if constants were codegen'ed into Python code. I think that's a `moonworm` issue, though and will take some work.